src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupLoggingHelper.cs

```diff
@@ -95,53 +93,93 @@ internal static string GetParameterText(string prefix, string parameterName, ILi
         /// </summary>
         internal static string GetStringFromParameterValue(object parameterValue)
         {
-            var type = parameterValue.GetType();
-
-            ErrorUtilities.VerifyThrow(!type.IsArray, "scalars only");
+            // fast path for the common case
+            if (parameterValue is string valueText)
+            {
+                return valueText;
+            }
 
-            if (type == typeof(string))
+            using (var sb = new ReuseableStringBuilder())
             {
-                return (string)parameterValue;
+                AppendStringFromParameterValue(sb, parameterValue);
+                return sb.ToString();
             }
-            else if (type.GetTypeInfo().IsValueType)
+        }
+
+        // Avoid allocating a temporary list to hold metadata for sorting every time.
+        // Each thread gets its own copy.
+        [ThreadStatic]
+        private static List<KeyValuePair<string, string>> keyValuePairList;
+
+        private static void AppendStringFromParameterValue(ReuseableStringBuilder sb, object parameterValue)
+        {
+            if (parameterValue is string text)
             {
-                return (string)Convert.ChangeType(parameterValue, typeof(string), CultureInfo.CurrentCulture);
+                sb.Append(text);
             }
-            else if (typeof(ITaskItem).GetTypeInfo().IsAssignableFrom(type.GetTypeInfo()))
+            else if (parameterValue is ITaskItem item)
             {
-                var item = ((ITaskItem)parameterValue);
-                string result = item.ItemSpec;
+                sb.Append(item.ItemSpec);
 
                 var customMetadata = item.CloneCustomMetadata();
+                int count = customMetadata.Count;
 
-                if (customMetadata.Count > 0)
+                if (count > 0)
                 {
-                    result += "\n";
-                    var names = new List<string>();
+                    sb.Append('\n');
+
+                    // need to initialize the thread static on each new thread
+                    if (keyValuePairList == null)
+                    {
+                        keyValuePairList = new List<KeyValuePair<string, string>>(count);
+                    }
 
-                    foreach (string name in customMetadata.Keys)
+                    var customMetadataDictionary = customMetadata as IDictionary<string, string>;
+                    if (customMetadataDictionary != null)
```

Comment:
nit: `if (customMetadata is IDictionary<string, string> customMetadataDictionary)`